quinn-proto/src/connection.rs

```diff
@@ -1031,58 +1037,87 @@ impl Connection {
         }
 
         trace!(ctx.log, "connection got packet"; "connection" => %self.loc_cid, "len" => packet.payload.len());
-        let was_closed = self.state.as_ref().unwrap().is_closed();
+        let (prev_state, was_handshake) = match self.state.take().unwrap() {
+            State::Handshake(mut state) => {
+                if !state.rem_cid_set {
+                    match packet.header {
+                        Header::Long {
+                            src_cid: rem_cid, ..
+                        } => {
+                            trace!(ctx.log, "got remote connection id"; "connection" => %self.loc_cid, "rem_cid" => %rem_cid);
+                            self.rem_cid = rem_cid;
+                            state.rem_cid_set = true;
+                        }
+                        _ => {}
+                    }
+                }
+                (State::Handshake(state), true)
+            }
+            state => (state, false),
+        };
+        let was_closed = prev_state.is_closed();
 
-        // State transitions
-        let prev_state = self.state.take().unwrap();
-        let was_handshake = match prev_state {
-            State::Handshake(_) => true,
-            _ => false,
+        let number = match self.decrypt_packet(was_handshake, &mut packet) {
+            Ok(number) => Some(number),
+            Err(Some(e)) => {
+                warn!(ctx.log, "got illegal packet"; "connection" => %self.loc_cid, "reason" => %e);
+                None
```

Comment:
These errors are fatal to a handshaking or established connection and must produce the appropriate state transition and close frame, as with errors from `handle_connected_inner` where this call was formerly made.

Reply:
Ah yeah, I was aware I had something fishy going on here.

Reply:
I think the new version of this commit (d26777df2414958acb5b479de4345a76acd5d24b) should be better, can you review that?

Reply:
Looks good!